langserver/symbol.go

```diff
@@ -239,17 +250,84 @@ func score(q Query, s lsp.SymbolInformation) (scor int) {
 
 // toSym returns a SymbolInformation value derived from values we get
 // from the Go parser and doc packages.
-func toSym(name, container string, kind lsp.SymbolKind, fs *token.FileSet, pos token.Pos) lsp.SymbolInformation {
-	container = filepath.Base(container)
-	if i := strings.LastIndex(container, " "); i >= 0 {
-		container = container[i+1:]
+func toSym(name string, bpkg *build.Package, recv string, kind lsp.SymbolKind, fs *token.FileSet, pos token.Pos) symbolPair {
+	container := recv
+	if container == "" {
+		container = filepath.Base(bpkg.ImportPath)
+	}
+	return symbolPair{
+		SymbolInformation: lsp.SymbolInformation{
+			Name:          name,
+			Kind:          kind,
+			Location:      goRangeToLSPLocation(fs, pos, pos+token.Pos(len(name))-1),
+			ContainerName: container,
+		},
+		// NOTE: fields must be kept in sync with workspace_refs.go:defSymbolDescriptor
+		desc: lspext.SymbolDescriptor{
+			"vendor":      IsVendorDir(bpkg.Dir),
+			"package":     path.Clean(bpkg.ImportPath),
+			"packageName": bpkg.Name,
+			"recv":        recv,
+			"name":        name,
+			"id":          fmt.Sprintf("%s:%s:%s:%s", path.Clean(bpkg.ImportPath), bpkg.Name, recv, name),
+		},
+	}
+}
+
+func parseID(v string) (pkg, packageName, recv, name string) {
```

Comment:
This is exactly the kind of thing we want to avoid with an `id` field. The IDs should only be compared on the basis of string equality. If we start encoding structure into the ID, we're basically trying to encode a `SymbolDescriptor` into a string format. In that case, we might as well use a standard encoding (like URL encoding). This also means we're treating `id` differently from other `SymbolDescriptor` fields, which is hacky and unexpected.